File: demo/KSYLiveDemo/KSYUIUtils/KSYNameSlider.m

#import "KSYNameSlider.h"

@implementation KSYNameSlider

- (id) init {
    self = [super init];
    UIColor* bgColor = [UIColor colorWithWhite:0.8 alpha:0.3];
    self.nameL  = [[UILabel alloc] init];
    self.valueL = [[UILabel alloc] init];
    self.nameL.backgroundColor  = bgColor;
    self.valueL.backgroundColor = bgColor;
    self.slider = [[UISlider alloc] init];
    self.slider.minimumValue = 0;
    self.slider.maximumValue = 100;
    [self addSubview:_nameL];
    [self addSubview:_valueL];
    [self addSubview:_slider];
    [_slider addTarget:self
                action:@selector(onSlider:)
      forControlEvents:UIControlEventValueChanged];
    self.valueL.textAlignment = NSTextAlignmentCenter;
    self.onSliderBlock = nil;
    _normalValue = (_slider.value - _slider.minimumValue) / _slider.maximumValue;
    _precision = 0;
    return self;
}
- (void)setFrame:(CGRect)frame{
    [super setFrame:frame];
    [self layoutSlider];
}
- (void)layoutSlider{
    CGFloat wdt = self.frame.size.width;
    CGFloat hgt = self.frame.size.height;
    [_nameL sizeToFit];
    [_valueL sizeToFit];
    
    CGFloat wdtN = _nameL.frame.size.width + 10;
    CGFloat wdtV = _valueL.frame.size.width + 10;
    CGFloat wdtS = wdt - wdtN - wdtV;
    _nameL.frame  = CGRectMake(0, 0, wdtN, hgt);
    _slider.frame = CGRectMake(wdtN, 0, wdtS, hgt);
    _valueL.frame = CGRectMake(wdtN+wdtS, 0,wdtV, hgt);
}
- (void) updateValue {
    float val = _slider.value;
    if (_precision == 0){
        _valueL.text = [NSString stringWithFormat:@"%d", (int)val];
    }
    else {
        NSString *fmt =[NSString stringWithFormat:@"%%0.%df", _precision];
        _valueL.text = [NSString stringWithFormat:fmt, val];
    }
    [self layoutSlider];
    _normalValue = (_slider.value - _slider.minimumValue) / _slider.maximumValue;
}

//UIControlEventValueChanged
- (IBAction)onSlider:(id)sender {
    [self updateValue];
    if (_onSliderBlock) {
        _onSliderBlock(self);
    }
}

@synthesize normalValue = _normalValue;
- (float)normalValue{
    return _normalValue;
}
- (void)setNormalValue:(float )val{
    _slider.value = val * _slider.maximumValue + _slider.minimumValue;
    [self updateValue];
}
- (float)value{
    return _slider.value;
}
- (void)setValue:(float )val{
    _slider.value = val;
    [self updateValue];
    if (_onSliderBlock) {
        _onSliderBlock(self);
    }
}
@end

File: demo/KSYLiveDemo_Swift/KSYUIUtils_Swift/KSYNameSlider.swift
//
//  KSYNameSlider.swift
//  KSYLiveDemo_Swift
//
//  Created by iVermisseDich on 17/1/9.
//  Copyright © 2017年 com.ksyun. All rights reserved.
//

import UIKit
/**
 自定义控件: 带标签的slider
 
 控件格式为: 参数名称 + 滑块 + 当前数值
 通过 onSliderBlock,可以得到滑块值改变的事件
 通过 normalValue, 可以查询到归一化的值(0~1.0)
 */

class KSYNameSlider: UIView {

    var nameL: UILabel     // 滑块对应参数名称
    var slider: UISlider   // 滑块
    var valueL: UILabel    // 滑块当前的值
    
    // UIControlEventValueChanged 回调
    var onSliderBlock: ((AnyObject) -> Void)?
    // normalize between 0.0~1.0 [ (value-min)/max ]
    var _normalValue: Float = 0
    var normalValue: Float {
        get{
            return _normalValue
        }
        set{
            _normalValue = newValue
            slider.value = newValue * slider.maximumValue + slider.minimumValue
            self.updateValue()
        }
    }
    
    // 值的小数点位数[0,1,2,3]
    var precision: Int = 0
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    init() {
        let bgColor: UIColor = UIColor.init(white: 0.8, alpha: 0.3)
        
        nameL = UILabel.init()
        valueL = UILabel.init()
        slider = UISlider.init()
        
        super.init(frame: CGRect.zero)
        
        nameL.backgroundColor = bgColor
        valueL.backgroundColor = bgColor
        valueL.textAlignment = .center
        
        slider.minimumValue = 0
        slider.maximumValue = 100
        
        addSubview(nameL)
        addSubview(valueL)
        addSubview(slider)
        
        slider.addTarget(self, action: #selector(onSlider), for: .valueChanged)
        onSliderBlock = nil
        normalValue = (slider.value - slider.minimumValue) / slider.maximumValue;
        precision = 0
    }
    
    override var frame: CGRect {
        get{
            return super.frame
        }
        set{
            super.frame = newValue
            layoutSlider()
        }
    }
    
    func layoutSlider() {
        let wdt = self.frame.width
        let hgt = self.frame.height
        
        nameL.sizeToFit()
        valueL.sizeToFit()

        let wdtN: CGFloat = nameL.frame.width + 10;
        let wdtV: CGFloat = valueL.frame.width + 10;
        let wdtS: CGFloat = wdt - wdtN - wdtV;
        nameL.frame  = CGRect.init(x: 0, y: 0, width: wdtN, height: hgt)
        slider.frame = CGRect.init(x: wdtN, y: 0, width: wdtS, height: hgt)
        valueL.frame = CGRect.init(x: wdtN+wdtS, y: 0, width: wdtV, height: hgt)
    }
    
    func updateValue() {
        let val = slider.value
        if precision == 0 {
            valueL.text = "\(Int(val))"
        }else{
            valueL.text = String.init(format: "%0.\(precision)f", val)
        }
        self.layoutSlider()
        _normalValue = (slider.value - slider.minimumValue) / slider.maximumValue
    }
    
    func onSlider() {
        self.updateValue()
        if (onSliderBlock != nil) {
            onSliderBlock!(self)
        }
    }
}
